File: project_euler/Problem032/Problem032.java
package project_euler;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Problem032 {
    public void solution() {
        boolean[] digits = new boolean[9];
        Set<Integer> products = new HashSet<>();
        for (int i = 1; i < 10; ++i) {
            for (int j = 10000 / i; j >= 1234; --j) {
                getProducts(i, j, digits, products);
            }
        }

        for (int i = 10; i < 100; ++i) {
            for (int j = 10000 / i; j >= 123; --j) {
                getProducts(i, j, digits, products);
            }
        }

        int result = 0;
        for (int product : products) {
            result += product;
        }
        System.out.println(result);
    }

    private void getProducts(int multiplicand, int multiplier, boolean[] digits, Set<Integer> products) {
        Arrays.fill(digits, false);
        if (isPandigital(multiplicand, digits) && isPandigital(multiplier, digits)) {
            int product = multiplicand * multiplier;
            if (isPandigital(product, digits)) {
                products.add(product);
            }
        }
    }

    private boolean isPandigital(int value, boolean[] digits) {
        while (0 != value) {
            int index = value % 10 - 1;
            if (index < 0 || digits[index]) {
                return false;
            }
            digits[index] = true;
            value /= 10;
        }
        return true;
    }
}

File: project_euler/Problem032/Problem032.py
from typing import List

class Problem032:
    def solution(self) -> None:
        digits, products = [False for _ in range(9)], set()
        for i in range(1, 10):
            for j in range(10000 // i, 1233, -1):
                self.__get_products(i, j, digits, products)

        for i in range(10, 100):
            for j in range(10000 // i, 122, -1):
                self.__get_products(i, j, digits, products)

        result = 0
        for product in products:
            result += product
        print(result)

    def __get_products(self, multiplicand: int, multiplier: int, digits: List[bool], products: set[int]):
        digits[:] = [False for _ in range(9)]
        if self.__is_pandigital(multiplicand, digits) and self.__is_pandigital(multiplier, digits):
            product = multiplicand * multiplier
            if self.__is_pandigital(product, digits):
                products.add(product)

    def __is_pandigital(self, value: int, digits: List[bool]) -> bool:
        while value:
            index = value % 10 - 1
            if index < 0 or digits[index]:
                return False
            digits[index] = True
            value //= 10
        return True
